File: tools/native/slots.c
/*
 * **Which of a routine's locals another routine reaches.**
 *
 * NOT a transcription. This exists because of a question the port cannot
 * answer from its own sources: a routine hands another the address of one of
 * its `[bp-N]` locals, and nothing static says how far the callee reads from
 * there. Get that wrong and a local that looks private is really element 2 of
 * an array whose base is a different local - which is exactly how
 * `draw_part_extra` lost two polygon corners.
 *
 * The measurement is only possible here. In the large model SS *is* DGROUP -
 * measured, 0x2e4c in every one of 40M sampled instructions, in every CS
 * including the overlays - so a near pointer to a local looks like a near
 * pointer to a global and the C cannot tell them apart. The emulator can: it
 * sees the address, and Borland's `push bp / mov bp,sp` leaves the frame chain
 * in memory to attribute it to.
 *
 * The rule for one access:
 *
 *   walk the BP chain outwards and find the innermost frame whose BP is above
 *   the address. Frame 0 is the routine doing the reading, so a hit there is
 *   its own local and is not interesting. A hit in frame k > 0 is a routine
 *   reading into a *caller's* frame, and `BP(k) - address` is which local.
 *
 * The routine owning frame k is named by the return address in frame k-1,
 * resolved the way `guest_backtrace` resolves one: the far reading and the
 * near reading are both tried against the symbol table and whichever lands in
 * a known routine wins. Where neither does the record still goes out, with the
 * raw addresses, because a run that names nothing is still evidence about how
 * deep the reach was.
 *
 * **Writes only, and it is the emulator that decides that.** `guest_mem` is
 * given to Unicorn with `uc_mem_map_ptr`, and over such a region a
 * `UC_HOOK_MEM_READ` that lets execution continue breaks the guest - with a
 * callback whose entire body is `return`, the game leaves the rails inside
 * twelve frames and the backtrace shows `game_main+0x8` calling
 * `part_flip_options`, which is not a call that exists.
 * `UC_HOOK_MEM_READ_AFTER` does the same; the identical write hook runs clean.
 * The one read hook already in `native.c`, `on_vga_access`, escapes it by
 * calling `uc_emu_stop` and never returning to running code.
 *
 * **And it sees only the code the emulator still runs.** A routine dispatched
 * to the port writes `guest_mem` from C and never goes near Unicorn, so its
 * accesses are invisible here. That is the opposite of a detail: the reaches
 * this was built to find are mostly in dispatched code - `draw_polygon`
 * walking three points from an address `draw_part_extra` handed it - and none
 * of those appear. What does appear is the still-emulated side, which is the
 * part no static reading of the port's sources can cover at all. Taking the
 * other half needs the same bookkeeping inside the port's own accessors, where
 * `dg_enter` already knows the frame.
 *
 * A routine whose prologue omits `push bp / mov bp,sp` has no frame and is not
 * in the chain, so a hit from inside it is attributed to whoever called it.
 * Borland builds a frame in everything that takes arguments, which is
 * everything that could have been handed an address, so this bites rarely -
 * but it is why the output says how deep a hit was rather than only who made
 * it.
 *
 * **Arguments look like a caller's locals and are not.** The four or six bytes
 * of linkage above a frame's BP are followed by the arguments the caller
 * pushed, and only above those do the caller's own locals begin. Nothing here
 * knows where the arguments end, so a hit close above BP is probably an
 * argument fetch. The distance is recorded and `slots.py` draws the line,
 * where it can be drawn against the port's own `dg_enter` sizes.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "native.h"
#include "../../reconstruct/dgroup.h"


static FILE    *g_out;
static uint32_t g_seen;

/*
 * Every (owner, local, reader) triple already written, in an open-addressed
 * table. A run makes millions of stack accesses and almost all of them repeat,
 * so without a dedupe the log is gigabytes of the same handful of facts - and
 * a *linear* dedupe is worse than none, because it turns a hook that fires on
 * every stack access into an O(n^2) scan. This hook is on the hot path of the
 * emulator and its cost is visible in how far the run gets.
 */
#define TRIP_SLOTS (1u << 16)
struct trip { uint32_t owner, reader; uint16_t off, depth; uint8_t used; };
static struct trip g_tab[TRIP_SLOTS];

static int32_t already(uint32_t owner, uint32_t reader, uint16_t off,
                       uint16_t depth)
{
    uint32_t h = owner * 2654435761u ^ reader * 40503u ^ (off << 3) ^ depth;
    uint32_t i;

    h &= TRIP_SLOTS - 1;
    for (i = 0; i < TRIP_SLOTS; i++) {
        struct trip *t = &g_tab[(h + i) & (TRIP_SLOTS - 1)];

        if (!t->used) {
            t->owner = owner; t->reader = reader;
            t->off = off; t->depth = depth; t->used = 1;
            return 0;
        }
        if (t->owner == owner && t->reader == reader
            && t->off == off && t->depth == depth)
            return 1;
    }
    return 1;                          /* full: stop recording, do not grow */
}

static uint16_t peek16(uint32_t linear)
{
    return (uint16_t)(guest_mem[linear] | (guest_mem[linear + 1] << 8));
}

/* The image offset a return address means, taking whichever reading lands. */
static uint32_t owner_of(uint16_t cs, uint16_t r_off, uint16_t r_seg)
{
    uint32_t far_lin  = (uint32_t)r_seg * 16 + r_off;
    uint32_t near_lin = (uint32_t)cs * 16 + r_off;

    if (far_lin >= IMAGE_BASE && sym_for(far_lin - IMAGE_BASE, NULL))
        return far_lin;
    if (near_lin >= IMAGE_BASE && sym_for(near_lin - IMAGE_BASE, NULL))
        return near_lin;
    return near_lin;                   /* unnamed; the address is still a fact */
}

void native_slots_open(const char *path)
{
    g_out = fopen(path, "w");
    if (g_out == NULL) {
        fprintf(stderr, "TIM_SLOTS: cannot write %s\n", path);
        return;
    }
    fprintf(g_out, "# owner reader local depth kind\n");
}

void native_slots_access(uc_engine *uc, uint32_t type, uint64_t address,
                         int32_t size, int64_t value, void *ud)
{
    uint16_t cs = 0, ip = 0, bp = 0, ss = 0;
    uint32_t base;
    uint16_t frames[24];
    int32_t n = 0, k;
    uint16_t off, owner_bp;
    uint32_t owner, reader;

    (void)size; (void)value; (void)ud;
    if (g_out == NULL)
        return;

    uc_reg_read(uc, UC_X86_REG_SS, &ss);
    base = (uint32_t)ss * 16;
    if (address < base || address >= base + 0x10000)
        return;
    off = (uint16_t)(address - base);

    uc_reg_read(uc, UC_X86_REG_BP, &bp);
    if (bp == 0)
        return;

    /*
     * **The fast path, and it is most of them.** An address below the current
     * BP is this routine's own local, which is the ordinary case and the one
     * this tool has nothing to say about. Deciding it costs two register reads
     * and a compare; only the minority at or above BP - arguments, and reaches
     * into a caller - pays for the chain walk below.
     */
    if (off < bp)
        return;

    /* The chain, outwards. It must climb; anything else is not a frame. */
    {
        uint16_t b = bp;

        while (n < 24) {
            uint16_t next;

            frames[n++] = b;
            next = peek16(base + b);
            if (next <= b)
                break;
            b = next;
        }
    }

    for (k = 0; k < n; k++)
        if (frames[k] > off)
            break;
    if (k <= 0 || k >= n)
        return;                        /* its own local, or off the chain */

    owner_bp = frames[k];
    uc_reg_read(uc, UC_X86_REG_CS, &cs);
    uc_reg_read(uc, UC_X86_REG_IP, &ip);
    owner = owner_of(cs, peek16(base + frames[k - 1] + 2),
                     peek16(base + frames[k - 1] + 4));
    reader = (uint32_t)cs * 16 + ip;

    if (already(owner, reader, (uint16_t)(owner_bp - off), (uint16_t)k))
        return;

    fprintf(g_out, "%07x %07x %u %d %c\n", owner, reader,
            (unsigned)(owner_bp - off), k,
            type == UC_MEM_WRITE ? 'w' : 'r');
    /* flushed as it goes: the run this most wants to watch is one that ends in
       a trap, and a buffered tail is exactly what such a run loses */
    fflush(g_out);
    g_seen++;
}

void native_slots_close(void)
{
    if (g_out == NULL)
        return;
    fclose(g_out);
    fprintf(stderr, "TIM_SLOTS: %u distinct (owner, reader, local) triples\n",
            g_seen);
    g_out = NULL;
}

File: tools/native/slots.py
#!/usr/bin/env python3
"""Read what `TIM_SLOTS` measured: whose locals another routine reaches.

The port models every `dg_enter` frame slot as a DGROUP offset, and the only
ones that have to be are those another routine is given the address of. Nothing
in the port's own sources says how far a callee reads from such an address - so
`draw_part_extra`'s `v02`, which is only ever `DG16(v02)`, is really `x[2]` of
an array whose base is the slot two along, and unpacking it into a C local left
a polygon reading garbage.

The emulator can answer it, because in the large model SS *is* DGROUP and
Borland leaves a frame chain in memory. `slots.c` records one line per distinct
(owner, reader, local) triple; this names them and says what it means for the
unpacking:

    routine F   [bp-N]   read by G, 1 frame down

Read the **distance** carefully, and note which way it runs. `slots.c` records
`ownerBP - address`, so the number is how far *below* the owner's BP the byte
is - which is exactly how the port already writes a local, `[bp-N]`. A routine
fetching its own arguments reads *above* its BP, which lands below the caller's
locals and so comes out as a distance larger than the caller's frame; and the
port states every frame's size in its own source, as the argument to
`dg_enter`. So the line between "a local of the caller" and "an argument or
something further out" is not a guess here: it is `dg_enter`'s number, read
from the routine the record names.

**It sees only what the emulator still runs.** A routine dispatched to the
port writes `guest_mem` from C and never passes through Unicorn, so nothing it
does appears here. The reaches that motivated this - `draw_polygon` walking
three points from an address `draw_part_extra` handed it - are all in
dispatched code and all invisible. What it does cover is the emulated side,
which is the part the port's own sources cannot be read for at all.

This file is the port's own tooling; it is not a transcription.
"""
import argparse
import collections
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(os.path.dirname(HERE))
sys.path.insert(0, HERE)
import gensyms
import glob
import re


def frame_sizes():
    """Each routine's frame, as the port's own `dg_enter` states it."""
    fn = re.compile(r"^[a-zA-Z_].*\b(\w+)\s*\(")
    out = {}
    for path in sorted(glob.glob(os.path.join(ROOT, "reconstruct", "src", "*.c"))):
        cur = None
        for line in open(path):
            m = fn.match(line)
            if m and not line.rstrip().endswith(";"):
                cur = m.group(1)
            m2 = re.search(r"dg_enter\((0x[0-9a-fA-F]+|\d+)\)", line)
            if m2 and cur:
                out[cur] = int(m2.group(1), 0)
    return out


def named(syms, addr, base):
    """The routine containing this linear address, and how far into it."""
    off = addr - base
    best = None
    for at in syms:
        if at <= off and (best is None or at > best):
            best = at
    if best is None or off - best > 0x2000:
        return None, off
    return syms[best][0], off - best


def main():
    ap = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("log", help="the file TIM_SLOTS wrote")
    ap.add_argument("--image-base", type=lambda s: int(s, 0), default=0x110 * 16,
                    help="where the program was loaded (default %(default)#x)")
    ap.add_argument("--beyond", action="store_true",
                    help="also list the hits that fall outside the owner's "
                         "own frame - arguments, or a frame this tool could "
                         "not size")
    ap.add_argument("--top", type=int, default=30)
    args = ap.parse_args()

    syms = gensyms.collect()
    sizes = frame_sizes()
    reach = collections.defaultdict(set)      # owner -> {(local, reader, depth)}
    beyond = collections.defaultdict(set)
    unnamed = 0

    for line in open(args.log):
        if line.startswith("#") or not line.strip():
            continue
        o, r, local, depth, kind = line.split()
        owner, _ = named(syms, int(o, 16), args.image_base)
        reader, _ = named(syms, int(r, 16), args.image_base)
        local, depth = int(local), int(depth)
        if owner is None:
            unnamed += 1
            continue
        where = reach if local <= sizes.get(owner, 0) else beyond
        where[owner].add((local, reader or "?", depth, kind))

    def show(title, table, note):
        print(title)
        print(note + "\n")
        rows = sorted(table.items(), key=lambda kv: -len(kv[1]))
        for owner, hits in rows[:args.top]:
            offs = sorted({h[0] for h in hits})
            who = sorted({h[1] for h in hits})
            print("  %-28s frame %#-6x %d: %s"
                  % (owner, sizes.get(owner, 0), len(offs),
                     " ".join("[bp-%#x]" % o for o in offs[:10])))
            print("  %-28s   by %s" % ("", ", ".join(who[:5])))
        print("  %d routines, %d triples\n"
              % (len(table), sum(len(v) for v in table.values())))

    show("ROUTINES WHOSE OWN LOCALS ANOTHER ROUTINE TOUCHES",
         reach,
         "  the distance is within the frame `dg_enter` reserves, so these are\n"
         "  that routine's `[bp-N]` locals and the slot cannot become a plain\n"
         "  C local while the reader exists.")
    if args.beyond:
        show("HITS OUTSIDE THE OWNER'S OWN FRAME", beyond,
             "  further below the owner's BP than its frame reaches: an\n"
             "  argument fetch, or a routine whose frame this could not size.")
    else:
        print("%d routines have hits outside their own frame (--beyond)"
              % len(beyond))
    if unnamed:
        print("%d records name no transcribed routine as the owner" % unnamed)
    return 0


if __name__ == "__main__":
    sys.exit(main())
